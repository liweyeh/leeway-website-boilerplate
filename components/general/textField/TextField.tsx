/**
 * TextField component
 * @author Liwei Yeh <liwei.yeh@adasasistemas.com>
 */

// Library
import { debounce } from 'lodash';

// CSS Style for this module
import styles from './textField.module.scss';

/**
 * Interface for TextField props
 */
interface TextFieldProps {
	/**
	 * Custom styles
	 */
	customStyles?: string;
	/**
	 * Preset type for the html input tag
	 */
	type?: 'text' | 'textarea';
	/**
	 * label of the input element
	 */
	label?: string;
	/**
	 * Default text that is displayed within the field
	 */
	placeholder?: string;
	/**
	 * Msg to be displayed when the status is error
	 */
	errorMsg?: string;
	/**
	 * String for determining whether a error should be displayed
	 */
	status?: 'valid' | 'error';
	/**
	 * Number for determining the text limit input
	 */
	maxLength?: number;
	/**
	 * onChange function for keeping track field value
	 * debounce is built in
	 */
	handleChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
	/**
	 * default value for edit mode
	 */
	defaultValue?: string;
}

/**
 * Text fields let users enter and edit text.
 * @param {TextFieldProps} Prop
 * @returns  an styled  input element
 */
export const TextField = ({
	customStyles,
	type = 'text',
	label = '',
	placeholder = 'Please input a new value',
	errorMsg = 'Please fill up the field',
	status = 'valid',
	maxLength,
	handleChange = () => {
		return;
	},
	defaultValue = '',
}: TextFieldProps) => {
	const TEXTAREA = 'textarea';
	const onType = () => {
		return debounce(e => handleChange(e), 10);
	};
	return (
		<div className={[styles.container, customStyles].join(' ')}>
			{label && (
				<label htmlFor={label}>
					<p>{label}</p>
				</label>
			)}
			{type === TEXTAREA ? (
				<textarea
					className={[styles.textArea, styles[`textArea-${status}`]].join(' ')}
					placeholder={placeholder}
					onChange={onType()}
				/>
			) : (
				<input
					name={label}
					className={[styles.textField, styles[`textField-${status}`]].join(' ')}
					type={type}
					placeholder={placeholder}
					defaultValue={defaultValue}
					onChange={onType()}
					maxLength={maxLength ? maxLength : undefined}
				/>
			)}
			<p
				className={[styles['error-msg'], status !== 'error' && styles['error-msg-hide']].join(' ')}
			>
				{errorMsg}
			</p>
		</div>
	);
};
