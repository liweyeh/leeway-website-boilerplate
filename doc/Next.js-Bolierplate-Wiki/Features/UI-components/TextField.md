[[_TOC_]]

# Overview

- Purpose: Text input UI element
- Type: General UI
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/textField)
- Created by: Liwei Yeh <liwei.yeh@adasasistemas.com>
- Created time: 25/08/2022

# Props

- customStyles<`string`>: Custom styles
- type<`'text' | 'textarea'`>: Pre-set type for the html input tag
- label<`string`>: Label of the input element
- placeholder<`string`>: Default text that is displayed within the field
- errorMsg<`string`>: Msg to be displayed when the status is error
- status<`'valid' | 'error'`>: String for determining whether a error should be displayed
- maxLength<`number`>: Number for determining the text limit input
- handleChange<`(event: React.ChangeEvent<HTMLInputElement>) => void`>: onChange function for keeping track field value; debounce is built in
- defaultValue<`string`>: default value for edit mode

# Usage

```jsx
<TextField customStyles={styles.text} status="valid" type="textarea" />
```
