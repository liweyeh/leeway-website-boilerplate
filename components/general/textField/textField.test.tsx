import { render } from '@testing-library/react';
import { TextField } from './TextField';
import '@testing-library/jest-dom';

const setup = (status: 'error' | 'valid') => {
	const { asFragment } = render(
		<TextField
			placeholder={'Please give it a name'}
			errorMsg="Please input a new name or press x to cancel"
			status={status}
			handleChange={() => {
				return;
			}}
		/>,
	);
	return { asFragment };
};

test('test textfield snapshot', () => {
	const { asFragment } = setup('valid');
	expect(asFragment()).toMatchSnapshot();
});

test('test error textfield snapshot', () => {
	const { asFragment } = setup('error');
	expect(asFragment()).toMatchSnapshot();
});
