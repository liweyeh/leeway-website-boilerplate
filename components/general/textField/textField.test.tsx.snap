// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test error textfield snapshot 1`] = `
<DocumentFragment>
  <div
    class="container "
  >
    <input
      class="textField textField-error"
      name=""
      placeholder="Please give it a name"
      type="text"
      value=""
    />
    <p
      class="error-msg false"
    >
      Please input a new name or press x to cancel
    </p>
  </div>
</DocumentFragment>
`;

exports[`test textfield snapshot 1`] = `
<DocumentFragment>
  <div
    class="container "
  >
    <input
      class="textField textField-valid"
      name=""
      placeholder="Please give it a name"
      type="text"
      value=""
    />
    <p
      class="error-msg error-msg-hide"
    >
      Please input a new name or press x to cancel
    </p>
  </div>
</DocumentFragment>
`;
